Proposes a tutorial on how to implement a manifold (#190)
JuliaManifolds/Manifolds.jl: docs/src/tutorials/manifold.md

For this tutorial it is helpful, if you heard of the exponential map, tangent vectors and the dimension of a manifold.

We start with two technical preliminaries. If you want to start directly, you can [skip](@ref manifold-tutorial-task) this paragraph and revisit it for two of the implementation details.

After that, we will

* [model](@ref manifold-tutorial-task) the manifold
* [implement](@ref manifold-tutorial-checks) two tests, so that points and tangent vectors can be checked for validity, for example also within [`ValidationManifold`](@ref),
* [implement](@ref manifold-tutorial-fn) two functions, the exponential map and the manifold dimension.

## [Technical preliminaries](@id manifold-tutorial-prel)

There is only two small  technical things we require.
First of all our [`Manifold`](@ref)`{𝔽}` has a parameter `𝔽`.
This parameter indicates the [`number_system`](@ref) the manifold is based on, e.g. real-valued (`ℝ`) or complex-valued (`ℂ`) data.

> This can be used to simultaneously implement both a real-valued and a complex-valued type of a manifold.

Author: If you have `Circle{R}` and `Circle{C}` that's easier to do a common implementation for some functions compared to introducing a common abstract type.

That might not have been, how we _got_ to that parametrization, but it's a good reason anyways.

> This can be used to simultaneously implement both a real-valued and a complex-valued type of a manifold.

Reviewer: Well, you can have a number system parameter in manifold types without having it as a parameter of `Manifold`. So a number system parameter of, let's say, `Symmetric` is there to avoid adding another abstract type while a number system parameter of the `Manifold` type it derives from is there to have `number_of_coordinates` for free.

`Circle` may not be the best example here because it overwrites `number_of_coordinates`.

> This can be used to simultaneously implement both a real-valued and a complex-valued type of a manifold.

Author: Having it as a parameter was a choice, yes. Still I think the `number_of_coordinates` is not be the main reasons for our manifolds (it was to pull it up to Manifold, yes).

In general our manifolds have this parameter for easy dispatch per field type. So what do you propose to write in the tutorial?

> This can be used to simultaneously implement both a real-valued and a complex-valued type of a manifold.

Reviewer: In a basic tutorial like this I think it would be enough to say that for real manifolds users need to put `ManifoldsBase.ℝ` there and not worry about it, and for complex manifolds tell the reader that it has important interactions with bases.

> This can be used to simultaneously implement both a real-valued and a complex-valued type of a manifold.

Author: So I could just leave out line 31? That would be fine with me.

> This can be used to simultaneously implement both a real-valued and a complex-valued type of a manifold.

Reviewer: I have added a suggested change below.

> This can be used to simultaneously implement both a real-valued and a complex-valued type of a manifold.

Author: Thanks for the careful review.

Let's look at this for an example. The exponential map $\exp_p\colon T_p\mathcal M \to \mathcal M$ that maps a tangent vector $X\in T_p\mathcal M$ from the tangent space at $p\in \mathcal M$ to the manifold.
The function [`exp`](@ref exp(M::Manifold, p, X)) has to know the manifold `M`, the point `p` and the tangent vector `X` as input, so to compute the resulting point `q` would be calling

```julia
q = exp(M, p, X)
```

If you already have allocated memory to the variable that should store the result, it is better to perform the computations directly in that memory and avoid reallocations. For example

```julia
q = similar(p)
exp!(M, q, p, X)
```

calls [`exp!`](@ref exp!(M::Manifold, q, p, X)), which modifies its input `q` and returns the resulting point in there.
Actually these two lines are the default implementation for [`exp`](@ref exp(M::Manifold, p, X)).

> Note that for a unified interface, the manifold `M` is _always_ the first parameter, and the variable the result will be stored to in the mutating variants is _always_ the second parameter.

Reviewer: There will be one exception to this rule, specifically when a function takes another function as an argument, and it would make sense to write that function using the do-notation. Then `M` would be second and the mutating variable would be third.

Author: But currently we don't have that.

Reviewer: Yes, we don't have that yet on the master branch but it's already a part of #110 . We can of course modify tutorial in that PR then.

Author: Maybe changing it, when necessary might be best. And it stays a mostly or usually, still.

Long story short: if possible, implement the mutating version [`exp!`](@ref exp!(M::Manifold, q, p, X)), you get the [`exp`](@ref exp(M::Manifold, p, X)) for free.
Many functions that build upon basic functions employ the mutating variant, too, to avoid reallocations.

## [Startup](@id manifold-tutorial-startup)

As a start, let's load `ManifoldsBase.jl` and import the functions we consider throughout this tutorial.
For implementing a manifold, loading the interface should suffice for quite some time.

```@example manifold-tutorial
using ManifoldsBase, LinearAlgebra, Test
import ManifoldsBase: check_manifold_point, check_tangent_vector, manifold_dimension, exp!
```

## [Goal](@id manifold-tutorial-task)

> As an example, let's implement the sphere, but with a radius $r$. Since this radius is a property inherent to the manifold.

Reviewer: The second sentence in this line is missing something.

The second information, we want to store is the dimension of the sphere, for example whether it's the 1-sphere, i.e. the circle, represented by vectors $p\in\mathbb R^2$ or the 2-sphere in $\mathbb R^3$.
Since the latter might be something we want to [dispatch](https://en.wikipedia.org/wiki/Multiple_dispatch) on, we model it as a parameter of the type. Hence we define

```@example manifold-tutorial
"""
    MySphere{N} <: Manifold{ℝ}

Define an `n`-sphere of radius `r`. Construct by `MySphere(radius,n)
"""
struct MySphere{N} <: Manifold{ManifoldsBase.ℝ} where {N}
    radius::Float64
end
MySphere(radius, n) = MySphere{n}(radius)
Base.show(io::IO, M::MySphere{n}) where {n} = print(io, "MySphere($(M.radius),$n)")
nothing #hide
```

Here, the last line just provides a nicer print of a variable of that type

> The manifold should only contain information that it might require independently of its points or tangents.

Reviewer: Might be worth adding that it should always be possible to determine the dimension of a manifold based on the manifold object only.

Author: Maybe I phrase that even further, when introducing the struct, which thing the struct should be able to do.

> Now we can already initialize our Manifold, we use later, the $2$-sphere of radius $1.5$.

Reviewer: ```suggestion
Now we can already initialize our manifold that we will use later, the $2$-sphere of radius $1.5$.
```

```@example manifold-tutorial

> S = MySphere(1.5,2)

Reviewer: ```suggestion
S = MySphere(1.5, 2)
```

```

## [Checking points and tangents](@id manifold-tutorial-checks)

If we have now a point, represented as an array, we would first like to check, that it is a valid point on the manifold.
For this one can use the easy interface [`is_manifold_point`](@ref is_manifold_point(M::Manifold, p; kwargs...)). This internally uses [`check_manifold_point`](@ref check_manifold_point(M, p; kwargs...)).
This is what we want to implement.

> We have to return the error, if `p` is not on `M` and `nothing` otherwise.

Reviewer: ```suggestion
We have to return the error if `p` is not on `M` and `nothing` otherwise.
```

We have to check two things: that a point `p` is a vector with `N+1` entries and it's norm is the desired radius.

> To spare a few lines, we can use [lazy evaluation](https://en.wikipedia.org/wiki/Lazy_evaluation) instead of `if` statements.

Reviewer: ```suggestion
To spare a few lines, we can use [short-circuit evaluation](https://docs.julialang.org/en/v1/manual/control-flow/#Short-Circuit-Evaluation-1) instead of `if` statements.
```

```

## [Functions on the manifold](@id manifold-tutorial-fn)

For the [`manifold_dimension`](@ref manifold_dimension(M::Manifold)) we have to just return the `N` parameter

```@example manifold-tutorial
manifold_dimension(::MySphere{N}) where {N} = N
manifold_dimension(S)
```

Note that we can even omit the variable name in the first line since we do not have to access any field or use the variable otherwise.

To implement the exponential map, we have to implement the formula for great arcs, given a start point `p` and a direction `X` on the $n$-sphere of radius $r$ the formula reads

````math
\exp_p X = \cos(\frac{1}{r}\lVert X \rVert)p + \sin(\frac{1}{r}\lVert X \rVert)\frac{r}{\lVert X \rVert}X.
````

Note that with this choice we for example implicitly assume a certain metric. This is completely fine. We only have to think about specifying a metric explicitly, when we have (at least) two different metrics on the same manifold.

An implementation of the mutation version, see the [technical note](@ref manifold-tutorial-prel), reads

```@example manifold-tutorial
function exp!(M::MySphere{N}, q, p, X) where {N}
    nX = norm(X)
    if nX == 0
        q .= p
    else

>         q.= cos(nX/M.radius)*p + M.radius*sin(nX/M.radius) .* (X./nX)

Reviewer: ```suggestion
        q .= cos(nX/M.radius)*p + M.radius*sin(nX/M.radius) .* (X./nX)
```

This tutorial demonstrates that you can get your first own manifold quite fast and you only have to implement the functions you actually need. For this tutorial it would be helpful if you take a look at our [notation](../notation.md).
This tutorial assumes that you heard of the exponential map, tangent vectors and the dimension of a manifold. If not, please read for example [[do Carmo, 1992](#doCarmo1992)],
Chapter 3, first.

We start with two technical preliminaries. If you want to start directly, you can [skip](@ref manifold-tutorial-task) this paragraph and revisit it for two of the implementation details.

After that, we will

* [model](@ref manifold-tutorial-task) the manifold
* [implement](@ref manifold-tutorial-checks) two tests, so that points and tangent vectors can be checked for validity, for example also within [`ValidationManifold`](@ref),
* [implement](@ref manifold-tutorial-fn) two functions, the exponential map and the manifold dimension.

## [Technical preliminaries](@id manifold-tutorial-prel)

There are only two small technical things we need to explain at this point.
First of all our [`Manifold`](@ref)`{𝔽}` has a parameter `𝔽`.

> This parameter indicates the [`number_system`](@ref) the manifold is based on, e.g. real-valued (`ℝ`) or complex-valued (`ℂ`) data.
> This can be used to simultaneously implement both a real-valued and a complex-valued type of a manifold.

Reviewer: ```suggestion
This parameter indicates the [`number_system`](@ref) the manifold is based on, for example `ℝ` for real manifolds. It is important primarily for defining bases of tangent spaces.
See [`Symmetric`](@ref) as an example of defining both a real-valued and a complex-valued symmetric manifolds using one type.
```
Maybe something like this?

Second, a main design decision of `Manifold.jl` is that most functions are implemented as mutating functions, i.e. as in-place-computations. There usually exists a non-mutating version that falls back to allocating memory and calling the mutating one. This means you only have to implement the mutating version, _unless_ there is a good reason to provide a special case for the non-mutating one, i.e. because in that case you know a far better performing implementation.

> Let's look at an example. The exponential map $\exp_p\colon T_p\mathcal M \to \mathcal M$ that maps a tangent vector $X\in T_p\mathcal M$ from the tangent space at $p\in \mathcal M$ to the manifold.

Reviewer: ```suggestion
Let's look at an example. The exponential map $\exp_p\colon T_p\mathcal M \to \mathcal M$ that maps a tangent vector $X\in T_p\mathcal M$ from the tangent space at $p\in \mathcal M$ to the manifold. You don't have to implement it for your manifold but it is very useful.
```

Author: What is the aim of this new sentence? It's anyways meant as an example, so I don't feel its necessary to state that it's a useful function (and sometimes you even can't implement it – in closed form).

Reviewer: Primarily I wouldn't want to leave the reader with an impression that the exponential map is necessary. Maybe it's poorly worded but at the moment it's somewhat unclear which functions are necessary for which functionality. Although that may be an idea for a different part of documentation. What do you think?

Author: You're right, that's a good point, but I'd like to add a paragraph in the introduction then; will think about one. I think here, also the place is not best, since its in the technical details section.

Reviewer: Good idea, a paragraph in the introduction would be better 👍 .

For implementing a manifold, loading the interface should suffice for quite some time.

```@example manifold-tutorial
using ManifoldsBase, LinearAlgebra, Test
import ManifoldsBase: check_manifold_point, check_tangent_vector, manifold_dimension, exp!
```

## [Goal](@id manifold-tutorial-task)

As an example, let's implement the sphere, but with a radius $r$.
Since this radius is a property inherent to the manifold, it will become a field of the manifold.
The second information, we want to store is the dimension of the sphere, for example whether it's the 1-sphere, i.e. the circle, represented by vectors $p\in\mathbb R^2$ or the 2-sphere in $\mathbb R^3$.
Since the latter might be something we want to [dispatch](https://en.wikipedia.org/wiki/Multiple_dispatch) on, we model it as a parameter of the type.

In general the `struct` of a manifold should provide information about the manifold, which are inherent to the manifold or has to be available without a specific point or tangent vector present.
This is -- most prominently -- a way to determine the manifold dimension.

For our example we define

```@example manifold-tutorial
"""
    MySphere{N} <: Manifold{ℝ}

> Define an `n`-sphere of radius `r`. Construct by `MySphere(radius,n)

Reviewer: ```suggestion
Define an `n`-sphere of radius `r`. Construct by `MySphere(radius,n)`
```

An implementation of the mutation version, see the [technical note](@ref manifold-tutorial-prel), reads

```@example manifold-tutorial
function exp!(M::MySphere{N}, q, p, X) where {N}
    nX = norm(X)
    if nX == 0
        q .= p
    else
        q .= cos(nX/M.radius)*p + M.radius*sin(nX/M.radius) .* (X./nX)
    end
    return q
end
nothing #hide
```

A first easy check can be done taking `p` from above and any vector `X` of length `1.5π` from its tangent space. The resulting point is opposite of `p`, i.e. `-p`

```@example manifold-tutorial
q = exp(S,p, [0.0,1.5π,0.0])
[isapprox(p,-q); is_manifold_point(S,q)]
```

## [Conclusion](@id manifold-tutorial-outlook)

You can now just continue implementing further functions from the [interface](../interface.md),
but with just [`exp!`](@ref exp!(M::Manifold, q, p, X)) you for example already have

* [`geodesic`](@ref geodesic(M::Manifold, p, X)) the (not necessarily shortest) geodesic emanating from `p` in direction `X`.

> * the [`ExponentialRetraction`](@ref), that the [`retract`](@ref retract(M::Manifold, p, X))ion uses by default.

Reviewer: ```suggestion
* the [`ExponentialRetraction`](@ref), that the [`retract`](@ref retract(M::Manifold, p, X)) function uses by default.
```

For the [`shortest_geodesic`](@ref shortest_geodesic(M::Manifold, p, q)) the implementation of a logarithm [`log`](@ref ManifoldsBase.log(M::Manifold, p, q)), again better a [`log!`](@ref log!(M::Manifold, X, p, q)) is necessary.

> Sometimes a default implementation is provided; for example if you implemented [`inner`](@ref inner(M::Manifold, p, X, Y)), the [`norm`](@ref norm(M, p, X)) is defined. You should overwrite it, if you can provide a more efficient version, gut for a start, the default should suffice.

Reviewer: ```suggestion
Sometimes a default implementation is provided; for example if you implemented [`inner`](@ref inner(M::Manifold, p, X, Y)), the [`norm`](@ref norm(M, p, X)) is defined. You should overwrite it, if you can provide a more efficient version. For a start the default should suffice.
```
